src/main/java/io/zold/api/Wallet.java

```diff
@@ -158,14 +187,54 @@ public void pay(final long amt, final long bnf) throws IOException {
             }
         }
 
-        // @todo #6:30min Implement merge method. This should merge this wallet
-        //  with a copy of the same wallet. It should throw an error if a
-        //  wallet is provided. Also add a unit test to replace
-        //  WalletTest.mergeIsNotYetImplemented().
         @Override
-        public Wallet merge(final Wallet other) {
-            throw new UnsupportedOperationException(
-                "merge() not yet supported"
+        // @todo #16:30min Following transactions should be ignored according
+        //  to the whitepaper:
+        //  a) If the transaction is negative and its signature is not valid,
+        //  it is ignored;
+        //  b) If the transaction makes the balance of the wallet negative,
+        //  it is ignored;
+        //  c) If the transaction is positive and it’s absent in the paying
+        //  wallet (which exists at the node), it’s ignored; If the paying
+        //  wallet doesn’t exist at the node, the transaction is ignored;
+        //
+        // @todo: #16:30min Merge method should update transactions
+        // in wallet's file and return concrete implementation not a fake one.
+        // Beware that tests should be refactored to take care of file cleanup
+        // after each case that merges wallets.
+        //
+        public Wallet merge(final Wallet other) throws IOException {
+            if (other.id() != this.id()) {
+                throw new IOException(
+                    new UncheckedText(
+                        new FormattedText(
+                            "Wallet ID mismatch, ours is %d, theirs is %d",
+                            other.id(),
+                            this.id()
+                        )
+                    ).asString()
+                );
+            }
+            final Iterable<Transaction> ledger = this.ledger();
+            final Iterable<Transaction> candidates = new Filtered<>(
+                origin -> new Filtered<>(
+                    incoming -> new UncheckedScalar<>(
```

Comment:
@Vatavuk wouldn't it make more sense if we switched the order of the names of `origin` and `incoming`? Because `origin` is in this case coming from `other.ledger()`

Reply:
@llorllale you are right, fixed